worker/rhpv3.go

```diff
@@ -16,6 +17,50 @@ import (
 	"go.sia.tech/renterd/api"
 )
 
+var (
+	// errBalanceMaxExceeded occurs when a deposit would push the account's
+	// balance over the maximum allowed ephemeral account balance.
+	errBalanceMaxExceeded = errors.New("ephemeral account maximum balance exceeded")
+)
+
+func (w *worker) fundAccount(ctx context.Context, account *account, pt rhpv3.HostPriceTable, hostIP string, hostKey types.PublicKey, amount types.Currency, revision *types.FileContractRevision) error {
+	return account.WithDeposit(ctx, func() (types.Currency, error) {
+		return amount, w.withTransportV3(ctx, hostIP, hostKey, func(t *rhpv3.Transport) (err error) {
+			rk := w.deriveRenterKey(hostKey)
+			cost := amount.Add(pt.FundAccountCost)
+			payment, ok := rhpv3.PayByContract(revision, cost, rhpv3.Account{}, rk) // no account needed for funding
+			if !ok {
+				return errors.New("insufficient funds")
+			}
+			w.contractSpendingRecorder.Record(revision.ParentID, api.ContractSpending{FundAccount: cost})
+			return RPCFundAccount(t, &payment, account.id, pt.UID)
+		})
+	})
+}
+
+func (w *worker) syncAccount(ctx context.Context, account *account, pt rhpv3.HostPriceTable, hostIP string, hostKey types.PublicKey) error {
```

Comment:
It fetches the balance from the host and then updates the local balance of the worker and the balance tracked by the bus. It needs to fetch the balance within `WithSync` to prevent any other operations from spending money in the meantime.